hashgraph/hedera-services
# 5636: MerkleDb may throw IllegalStateException when taking a snapshot after reconnect

Fix summary:
* A new concept is introduced to MerkleDb: primary and secondary tables. Primary tables are long-lived, they are used by VirtualMaps that a part of the merkle tree, they are included to snapshots and DB metadata. Secondary tables are short-lived, they are used during reconnects (both on the teacher and the learner sides). Once the corresponding data source is closed, secondary tables are deleted, both data and metadata
* During snapshots, only primary tables are included
* A new method is added to `VirtualDataSourceBuilder`: `copy()` It copies the given data source and, if requested, makes the copy primary. The method is used in `VirtualRootNode`. When a root node is detached for snapshots, the original data source is left unchanged (primary). During reconnects, the copy is marked as primary, while the original data source is changed to secondary. Previously, `snapshot()` method was used in these places, but its semantics is slightly different
* At HS level, a single MerkleDb instance is used in `ServicesState`. Previously, different instances could be used in some cases
* Some HS tests rely on the underlying data sources (JasperDB or MerkleDb) to be different for every test. Now with a single MerkleDb instance always used, `ServicesStateTest` is updated accordingly

Testing:
* All existing unit tests in hedera-node and platform-sdk
* A couple new unit tests are added
* Manual testing:
  * Start 3 HS nodes
  * Force a snapshot
  * Shut down the nodes
  * For one of the nodes, delete the snapshot to force it to reconnect on startup
  * Start the nodes again
  * Make sure the node falls behind and then reconnects
  * Check MerkleDb folders on disk
  * Check logs for exceptions

Fixes: https://github.com/hashgraph/hedera-services/issues/5636
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>
